File: Roost/backend/src/main/java/com/travelplanner/roost/common/error/ErrorCode.java
package com.travelplanner.roost.common.error;

import org.springframework.http.HttpStatus;

public enum ErrorCode {
    VALIDATION_ERROR("validation_error", HttpStatus.BAD_REQUEST),
    UNAUTHORIZED("unauthorized", HttpStatus.UNAUTHORIZED),
    FORBIDDEN("forbidden", HttpStatus.FORBIDDEN),
    NOT_FOUND("not_found", HttpStatus.NOT_FOUND),
    CONFLICT("conflict", HttpStatus.CONFLICT),
    THROTTLED("throttled", HttpStatus.TOO_MANY_REQUESTS),
    PAYMENT_FAILED("payment_failed", HttpStatus.PAYMENT_REQUIRED),
    INTERNAL_ERROR("internal_error", HttpStatus.INTERNAL_SERVER_ERROR);

    private final String code;
    private final HttpStatus status;

    ErrorCode(String code, HttpStatus status) {
        this.code = code;
        this.status = status;
    }

    public String code() {
        return code;
    }

    public HttpStatus status() {
        return status;
    }
}

File: TripWeaver/backend/Common/Errors/ErrorCode.cs
namespace TripWeaver.Common.Errors;

public enum ErrorCode
{
    ValidationError,
    Unauthorized,
    Forbidden,
    NotFound,
    Conflict,
    Throttled,
    InternalError,
}

public static class ErrorCodeExtensions
{
    public static string Code(this ErrorCode code) => code switch
    {
        ErrorCode.ValidationError => "validation_error",
        ErrorCode.Unauthorized => "unauthorized",
        ErrorCode.Forbidden => "forbidden",
        ErrorCode.NotFound => "not_found",
        ErrorCode.Conflict => "conflict",
        ErrorCode.Throttled => "throttled",
        _ => "internal_error",
    };

    public static int Status(this ErrorCode code) => code switch
    {
        ErrorCode.ValidationError => StatusCodes.Status400BadRequest,
        ErrorCode.Unauthorized => StatusCodes.Status401Unauthorized,
        ErrorCode.Forbidden => StatusCodes.Status403Forbidden,
        ErrorCode.NotFound => StatusCodes.Status404NotFound,
        ErrorCode.Conflict => StatusCodes.Status409Conflict,
        ErrorCode.Throttled => StatusCodes.Status429TooManyRequests,
        _ => StatusCodes.Status500InternalServerError,
    };
}
